_posts/2021-03-09-bezier-fitting.md

```diff
@@ -0,0 +1,196 @@
+
+Two things more to say: for smooth curves + low error, leash is generally perpendicular. Also, normalized arc length (both engineer and Pierre move at a constant speed) is a good approximation, though strictly larger.
+
+## Stating the problem
+
+Our dog Pierre has a funny quirk: he is capable of moving only along a cubic bezier path. That said, he is otherwise fairly obedient, and always starts and ends at the same position as the engineer, as well as starting out and ending up in the same direction.
+
+Another way to state this is that our cubic bezier must match the source curve to G1. Then we want to find values for the parameters that minimizes the error norm. For a cubic bezier, the two parameters are the lengths of the two control arms; all other parameters are determined by the requirement to match the source curve.
+
+As a mathematical simplification, we can factor out translation, uniform scaling, and rotation, and just consider a curve that goes from (0, 0) to (1, 0), with given $\theta_0$ and $\theta_1$ angles.
+
+[image]
+
+Then, the control points of the bezier are $(0, 0)$, $(\delta_0 \cos \theta_0, \delta_0 \sin \theta_0)$, $(1 - \delta_1 \cos \theta_1, \delta_1 \sin \theta_1)$, $(1, 0)$.
+
+## Prior work
+
+Having stated the problem, we can take a look at the existing literature.
+
+* G2 geometric Hermite interpolation, de Boor. A bunch of new things here: a proof of O(n^6) scaling. The presence of up to three solutions. However, the constant factor on error is 5-10x, which is not great.
+
+* Two approaches from the Penner paper, center of mass and orthogonal distance fitting (ODF).
+
+* The Graphics Gems chapter.
+
+Also note, I addressed this problem in chapter 9 of my [thesis]. That approach gave good results but was very slow. The solution in this blog is strictly better.
+
+## Solving the problem
+
+We have two parameters to optimize. A good place to start is a heat map of the error over the entire two dimensional parameter space. Below is such a heatmap for the Euler spiral with deflections 0.31 and 0.35 radians at the endpoints. The horizontal axis is $\delta_0$, ie the length of the left control arm, and the vertical axis is $\delta_1$, the length of the right control arm.
+
+![A 2D heatmap of error as a function of arm lengths](/assets/bezier_heatmap.png)
+
+Looking at this heatmap, we immediately see two things. First, there is clearly a curved line through the parameter space where the error is lower than on either side. On closer examination, the exact curve is an axis-aligned hyperbola, and we'll be able to characterize it soon. Second, along this curve the story is more complicated. In fact, for this source curve there are three local minima.
+
```

Comment:
It would be nice if this image had increased contrast or had an additional color for absolute lower bound of the range, because you have to really squint to see the minima.

Another option might be to just circle them or something?

Reply:
I've experimented with the colormap and upped the contrast basically as much as I could. I might fine-tune it some more, but because it's shown very clearly in the following 1D plot, I might consider it "good enough."

Reply:
Hm, I'd still probably circle, or arrow, the minima because between bezier_heatmap.png and bezier_heatmap_area.png, I straight up can't see the minima we're talking about.

Reply:
Ok, have a new version with arrows. Thanks for the feedback :)